### PR TITLE
[evm] enable Random opcode (EIP-4399) at Redsea height

### PR DESCRIPTION
**Pull request by dustinxie**
# Description
Please include a summary of the changes and the related issue. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes #(issue)

## Type of change
Please delete options that are not relevant.
- [] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [] Code refactor or improvement
- [] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [x] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: action/protocol/execution/evm/evm.go [hunk 1]
```diff
@@ -418,6 +418,12 @@ func executeInEVM(ctx context.Context, evmParams *Params, stateDB *StateDBAdapte
 		config = vmCfg
 	}
 	chainConfig := getChainConfig(g, blockHeight, evmParams.evmNetworkID)
+	if g.IsRedsea(blockHeight) {
+		// Merge enabled at Redsea height
+		tipHash := protocol.MustGetBlockchainCtx(ctx).Tip.Hash
+		h := common.BytesToHash(tipHash[:])
+		evmParams.context.Random = &h
```
Comment (envestcc) [inline action/protocol/execution/evm/evm.go:425]: The initialization of the `Random` field should occur in where `evmParams` is created.

File: action/protocol/execution/protocol_test.go [hunk 2]
```diff
@@ -523,7 +524,7 @@ func (sct *SmartContractTest) deployContracts(
 func (sct *SmartContractTest) run(r *require.Assertions) {
 	// prepare blockchain
 	ctx := context.Background()
-	cfg := config.Default
+	cfg := deepcopy.Copy(config.Default).(config.Config)
```
Comment (dustinxie) [inline action/protocol/execution/protocol_test.go:527]: I will add a `config.Clone()` func to remove this import/dependency

File: action/protocol/execution/evm/evm.go [hunk 3]
```diff
@@ -418,6 +418,12 @@ func executeInEVM(ctx context.Context, evmParams *Params, stateDB *StateDBAdapte
 		config = vmCfg
 	}
 	chainConfig := getChainConfig(g, blockHeight, evmParams.evmNetworkID)
+	if g.IsRedsea(blockHeight) {
+		// Merge enabled at Redsea height
+		tipHash := protocol.MustGetBlockchainCtx(ctx).Tip.Hash
```
Comment (Liuhaai) [inline action/protocol/execution/evm/evm.go:423]: TIP.Hash is not of randomness. Here is ETH's design for reference: https://github.com/ethereum/consensus-specs/blob/v1.4.0-beta.3/specs/phase0/beacon-chain.md#randao

**Review by Liuhaai — CHANGES_REQUESTED** [review]

File: action/protocol/execution/evm/evm.go (hunk 1 above)
Comment (dustinxie) [reply action/protocol/execution/evm/evm.go]: moved to `newParam`

File: action/protocol/execution/evm/evm.go (hunk 3 above)
Comment (dustinxie) [reply action/protocol/execution/evm/evm.go]: yes, in Beacon chain validator's BLS signature is used as the contribution to entropy, and its hash is XORed into the RANDAO during the update. The Beacon chain maintains the states, and the randomness is reflected as `MixDigest   common.Hash` in Ethereum's block header.

Currently we don't have these in place nor the corresponding data field in our block header.  We could implement a similar scheme but that would take quite some time.

File: action/protocol/execution/evm/evm.go [hunk 4]
```diff
@@ -432,7 +438,7 @@ func executeInEVM(ctx context.Context, evmParams *Params, stateDB *StateDBAdapte
 	remainingGas -= intriGas
 
 	// Set up the initial access list
-	rules := chainConfig.Rules(evm.Context.BlockNumber, g.IsRedsea(blockHeight)) // Merge enabled at Redsea height
+	rules := chainConfig.Rules(evm.Context.BlockNumber, evm.Context.Random != nil)
```
[the remark this answers is not in the document]
Comment (dustinxie) [reply action/protocol/execution/evm/evm.go:445]: recently I feel more and more that align with Ethereum code is beneficial to our codebase maintenance in the long run, that's why I changed to `evm.Context.Random != nil`

File: action/protocol/execution/evm/evm.go (hunk 3 above)
Comment (envestcc) [reply action/protocol/execution/evm/evm.go]: for evm, `prevrandao()` instruction always return a same value (i.e. equals to Random field of BlockCtx) during one block, according to implementation of ethereum. So it's not randomness during one block.

**Comment by dustinxie** [comment]
EIP-4399 to be enabled in future release

**Pull request closed without merging**